File: src/main/java/com/networkedassets/plugins/addpagetree/JsonPage.java
package com.networkedassets.plugins.addpagetree;

import com.atlassian.confluence.pages.Page;
import org.codehaus.jackson.annotate.JsonIgnoreProperties;
import org.codehaus.jackson.annotate.JsonProperty;

import java.util.List;
import java.util.stream.Collectors;

@SuppressWarnings("WeakerAccess") // has to be public for jackson to be happy
@JsonIgnoreProperties(ignoreUnknown = true)
class JsonPage {
    public String id;
    public String text;
    public List<JsonPage> children;
    @JsonProperty("a_attr")
    public Attr attr;
    public String icon = "icon-page";

    public static JsonPage from(Page page) {
        JsonPage jpage = new JsonPage();
        jpage.id = Long.toString(page.getId());
        jpage.text = page.getDisplayTitle();
        jpage.children = page.getSortedChildren().stream().map(JsonPage::from).collect(Collectors.toList());
        jpage.attr = new Attr();
        return jpage;
    }

    @JsonIgnoreProperties(ignoreUnknown = true)
    private static class Attr {
        @JsonProperty("data_added")
        public boolean isAdded = false;
    }
}

File: src/main/kotlin/com/networkedassets/plugins/managemytree/JsonPage.kt
package com.networkedassets.plugins.managemytree

import com.atlassian.confluence.pages.Page
import com.atlassian.confluence.security.Permission
import com.atlassian.confluence.security.PermissionManager
import com.atlassian.confluence.user.ConfluenceUser
import org.codehaus.jackson.annotate.JsonIgnoreProperties

@Suppress("unused")
@JsonIgnoreProperties(ignoreUnknown = true)
class JsonPage(
        val id: String,
        val text: String,
        val children: List<JsonPage>,
        val type: String = "default",
        val a_attr: Attr
) {
    data class Attr(
            val data_canEdit: Boolean = false,
            val data_canRemove: Boolean = false
    )

    companion object {
        fun from(page: Page, user: ConfluenceUser, permissionManager: PermissionManager, isRoot: Boolean = false): JsonPage = JsonPage(
                id = page.id.toString(),
                text = page.displayTitle,
                a_attr = Attr(
                        permissionManager.hasPermission(user, Permission.EDIT, page),
                        if (isRoot) false else permissionManager.hasPermission(user, Permission.REMOVE, page)
                ),
                children = page.sortedChildren
                        .filter({ p -> permissionManager.hasPermission(user, Permission.VIEW, p) })
                        .map({ p -> JsonPage.from(p, user, permissionManager) })
        )
    }
}
